src/Database/LSMTree/Internal/Lookup.hs

```diff
@@ -1,32 +1,166 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE DeriveFunctor       #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
 module Database.LSMTree.Internal.Lookup (
-    Run
+    -- * Lookup preparation
+    RunLookupView (..)
+  , RunIx
+  , KeyIx
   , prepLookups
+  , bloomQueries
+  , indexSearches
   ) where
 
-import           Data.Foldable (Foldable (..))
+import           Control.Monad.Primitive
+import           Control.Monad.ST.Strict
+import           Data.Primitive.ByteArray
+import           Data.Primitive.PrimVar
+import qualified Data.Vector as V
+import qualified Data.Vector.Mutable as VM
+import qualified Data.Vector.Unboxed as VU
+import qualified Data.Vector.Unboxed.Mutable as VUM
 import           Database.LSMTree.Internal.Run.BloomFilter (Bloom)
 import qualified Database.LSMTree.Internal.Run.BloomFilter as Bloom
 import           Database.LSMTree.Internal.Run.Index.Compact (CompactIndex,
-                     PageSpan)
+                     PageSpan (..))
 import qualified Database.LSMTree.Internal.Run.Index.Compact as Index
 import           Database.LSMTree.Internal.Serialise
+import           System.FS.API (Handle)
+import           System.FS.BlockIO.API
+
+-- | View of a 'Run' with only the essentials for lookups.
+data RunLookupView h = RunLookupView {
+    rlvKOpsHandle :: !h
+  , rlvBloom      :: !(Bloom SerialisedKey)
+  , rlvIndex      :: !CompactIndex
+  }
+  deriving Functor
 
--- | TODO: placeholder type for a run, replace by actual type once implemented
-type Run fd = (fd, Bloom SerialisedKey, CompactIndex)
+-- | Prepare disk lookups by doing bloom filter queries, index searches and
+-- creating 'IOOp's. The result is a vector of 'IOOp's and a vector of indexes,
+-- both of which are the same length. The indexes record the run and key
+-- associated with each 'IOOp'.
+prepLookups ::
+     PrimMonad m
+  => V.Vector (RunLookupView (Handle h))
```

Comment:
> I suspect using Run here would make testing more complicated? But maybe we can restructure `Run` a bit to group a few fields into something like `RunLookupView`, which is then only allocated once at run construction?

Indeed, it would mean that tests have to have access to a full `Run`, which is probably not what we want

My idea for getting around this is to abstract over the run type and provide a class with appropriate specialize constraints. This would hopefully also avoid having to map over the run vector and allocate new structures!

Reply:
> We could of course also re-use this vector of run views across lookups (until a new run is added or a merge completes), but that doesn't seem appealing, with the redundant information that can potentlally get out of sync.

I was expecting that we would do something like this. That is we would cache things needed by the higher-performance bits of the code, especially the bloom filters. The way to avoid it getting out of sync is to treat it like a cache: discard and re-create it every time the runs/levels change.

Reply:
I'll add this in a follow-up PR